Accept functions to flow_run_name and task_run_name settings

rsampaths16 opened the pull request:
<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.

Happy engineering!
-->
Currently flow_run_name and task_run_name settings only accept template strings to generate new name. This PR enhances this feature by accept a function that expects input parameters dict and returns a string. This should ideally allow injecting flow run context and task run context too.
<!-- Include an overview here -->

### Example
<!-- 
Share an example of the change in action.

A code blurb is best. Changes to features should include an example that is executable by a new user.
If changing documentation, a link to a preview of the page is great.
 -->
 
#### Flow run name setting
 ```python
import datetime
from prefect import flow

def generate_flow_run_name(params):
    date = datetime.datetime.utcnow()
    name = "anon" if "name" not in params else str(params["name"])

    return f"{name}-on-{date:%A}"

@flow(flow_run_name=generate_flow_run_name)
def my_flow(name: str):
    pass

# creates a flow run called 'marvin-on-Thursday'
my_flow(name="marvin")
```

#### Task run name setting
```python
import datetime
from prefect import flow, task

def generate_task_name(params):
    name = "anon" if "name" not in params else params["name"]
    date = datetime.datetime.utcnow()

    return f"hello-{name}-on-{date:%A}")

@task(name="My Example Task", 
      description="An example task for a tutorial.",
      task_run_name=generate_task_name)
def my_task(name):
    pass

@flow
def my_flow():
    # creates a run with a name like "hello-marvin-on-Thursday"
    my_task(name="marvin")
```

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [X] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  <!-- If you do not have permission to add a label, a maintainer will add one for you -->

closes https://github.com/PrefectHQ/prefect/issues/8895

rsampaths16 commented on src/prefect/engine.py:
```diff
@@ -618,7 +618,14 @@ async def orchestrate_flow_run(
     if flow.flow_run_name:
-        flow_run_name = flow.flow_run_name.format(**parameters)
+        if callable(flow.flow_run_name):
+            flow_run_name = flow.flow_run_name(parameters)
```
**WARNING**: The parameters is sent directly to the `flow_run_name` function which could leave it vulnerable to mutation. We could try to send a shallow copy of the parameters dict, but still the individual args themselves could be mutated ( similar to how it's vulnerable in `str.format(**parameters)` ).

I'll leave the call on what to do here to the Prefect team.

madkinsz replied:
I think that's okay with me for now 👍

rsampaths16 commented on src/prefect/engine.py:
```diff
@@ -618,7 +618,14 @@ async def orchestrate_flow_run(
+        else:
+            raise RuntimeError(
+                "flow_run_name parameter is neither a str nor a callable"
+            )
```
In both `flow_run_name` and `task_run_name` I'm raising exceptions when they are neither `str` nor `callable`. Should this assertion be removed? If not should an assertion be added in the actual `Flow` and `Task` classes too?

madkinsz replied:
Raising errors at the flow/task class level is better — then we will help people do the right thing when writing their task / flow rather than at runtime.

rsampaths16 replied:
I'll make changes to raise `TypeError`'s in `Flow` & `Task` class level. I'm also thinking to leave the `RuntimeError`'s here. Please confirm if the same if acceptable.

madkinsz replied:
It's okay here as well, but we'll probably want to move this into a separate function since it's kind of long now.

rsampaths16 commented:
> Thanks for contributing!
> 
> Before we dive into a review of the implementation, let's chat about the interface a little. I'm thinking that since I've added `parameters` to the `prefect.runtime` module perhaps this should be `Callable[[], str]` and you should just pull data from the runtime as needed. What do you think?

I believe this would be a good approach. We won't force a people to accept parameters if they don't want to use it, and since they'd be available in `prefect.runtime` those who want them are not blocked.

madkinsz commented:
@rsampaths16 sounds good to me!

rsampaths16 commented:
Hey @madkinsz during testing I've realized that `prefect.runtime.flow_run` tries to access the parent's `flow_run_context`, making the current flows parameters inaccessible to it. I think we'll have to make the function accept parameters as a kwargs dictionary. What are your thoughts on this?

P.S: PR seems to be accidentally closed as I've force pushed a rebase. Will re-open again when I've completed the implementation.

rsampaths16 commented:
Hey @madkinsz - regarding https://github.com/PrefectHQ/prefect/issues/8895#issuecomment-1483060367 what would be the plan to address this? Since the `run_name` functions are being called before we are inside the context of the `flow/task`? I'm assuming other functions that are accepted wouldn't have context too.

rsampaths16 commented:
Hi @madkinsz !!

https://github.com/PrefectHQ/prefect/blob/c0828fa478f242610c4fe8dac87c05e364cb5094/src/prefect/engine.py#L1520-L1530


Upon going through the code above, is my understanding correct that since the `task_run_name` function will be called within `task_run_context` context, it'll be able to access the `prefect.runtime` parameters?

If yes, I'll try out accessing `prefect.runtime` from the `task_run_name` and check if a similar pattern works for flow too. Maybe that'd be a separate pull-request.

Do let me know if I'm looking at the wrong pattern here.

rsampaths16 commented:
Awesome @madkinsz! I'll test this in the morning see if I can raise a new revision where functions don't accept parameters as discussed earlier. Thanks.

rsampaths16 commented:
Hey @madkinsz I've made the discussed changes.
The summary is:
1. refactored few tests in flows and tasks by grouping them together
2. add flow_run_name and task_run_name function support: they don't accept any arguments.
3. move flow_run_name generation within the flow run context: so it can access parameters from `prefect.runtime.flow_run.parameters`

For more detailed info you can check-out the commit messages. :)

rsampaths16 commented:
Hey @madkinsz - bumping again incase you've missed this;

madkinsz commented on docs/tutorials/flow-task-config.md:
````diff
@@ -94,6 +94,26 @@ def my_flow(name: str, date: datetime.datetime):
 my_flow(name="marvin", date=datetime.datetime.utcnow())
 ```
 
+Additionally this setting also accepts a function that returns a string for the flow_run_name
````
```suggestion
Additionally this setting also accepts a function that returns a string for the flow run name:
```

rsampaths16 replied:
done

madkinsz commented on docs/tutorials/flow-task-config.md:
````diff
@@ -94,6 +94,26 @@ def my_flow(name: str, date: datetime.datetime):
+# creates a flow run called 'Thursday-is-a-nice-day'
+my_flow(name="marvin")
+```
+
````
Can we add a blurb here that says something like

> If you need access to information about the flow, use the `prefect.runtime` module. For example:

Then include an example that pulls the name of the flow and a parameter from the runtime?

rsampaths16 replied:
done

madkinsz commented on docs/tutorials/flow-task-config.md:
````diff
@@ -147,6 +167,29 @@ def my_flow():
+    # creates a run with a name like "Thursday-is-a-lovely-day"
+    my_task(name="marvin")
+```
+
````
Same as above re an additional example but with the task run runtime.

rsampaths16 replied:
done

madkinsz commented on src/prefect/engine.py:
```diff
@@ -633,6 +626,21 @@ async def orchestrate_flow_run(
                 client=client,
                 parameters=parameters,
             ) as flow_run_context:
+                # update flow run name
+                if flow.flow_run_name:
+                    flow_run_name = _resolve_custom_flow_run_name(
+                        flow=flow, parameters=parameters
+                    )
+
+                    await client.update_flow_run(
+                        flow_run_id=flow_run.id, name=flow_run_name
+                    )
+                    logger.extra["flow_run_name"] = flow_run_name
+                    logger.debug(
+                        f"Renamed flow run {flow_run.name!r} to {flow_run_name!r}"
+                    )
+                    flow_run.name = flow_run_name
```
This will update the name on each retry, right? We'll want to avoid setting the name more than once.

rsampaths16 replied:
This is updating on each retry. Made changes to avoid setting it more than once. Used the same pattern as implemented in task_run_name. Have also added relevant tests.

rsampaths16 replied:
done

madkinsz commented on src/prefect/flows.py:
```diff
@@ -148,7 +149,12 @@ def __init__(
+
+        if flow_run_name is not None:
+            if not isinstance(flow_run_name, str) and not callable(flow_run_name):
+                raise TypeError("'flow_run_name' is neither a string nor a callable")
```
Generally we want to give more information to help debug
```suggestion
                raise TypeError(f"Expected string or callable for 'flow_run_name'; got {type(flow_run_name).__name__}.")
```

rsampaths16 replied:
done

madkinsz commented on src/prefect/tasks.py:
```diff
@@ -214,7 +215,11 @@ def __init__(
 
+        if task_run_name is not None:
+            if not isinstance(task_run_name, str) and not callable(task_run_name):
+                raise TypeError("'task_run_name' is neither a string nor a callable")
```
See flow run name error comment

rsampaths16 replied:
done

madkinsz commented on src/prefect/engine.py:
```diff
@@ -2056,6 +2065,32 @@ def should_log_prints(flow_or_task: Union[Flow, Task]) -> bool:
+    if callable(task.task_run_name):
+        task_run_name = task.task_run_name()
+        if not isinstance(task_run_name, str):
+            raise RuntimeError("callable 'task_run_name' did not return a string")
```
We probably want something more verbose for debugging

```suggestion
            raise RuntimeError(f"Callable {task.task_run_name} for 'task_run_name' returned type {type(task_run_name).__name__} but a string is required.")
```

Similar for the other error messages here

rsampaths16 replied:
done

madkinsz commented on src/prefect/engine.py:
```diff
@@ -2056,6 +2065,32 @@ def should_log_prints(flow_or_task: Union[Flow, Task]) -> bool:
+    if callable(flow.flow_run_name):
+        flow_run_name = flow.flow_run_name()
+        if not isinstance(flow_run_name, str):
+            raise RuntimeError("callable 'flow_run_name' did not return a string")
```
Note all of these `RuntimeError`s can be `TypeError`s

rsampaths16 replied:
Hey, I'll convert `RuntimeError`s to `TypeError`s where we check the type of the generating function/string. I'm not sure `TypeError` should be raised for the return value of the function.

rsampaths16 replied:
Just checked that `ValueError` is being raised on invalid return values in the code-base. Since the practice is established, I'll raise `TypeError` for the non string return value.

rsampaths16 replied:
done

madkinsz commented:
Thanks again for contributing @rsampaths16! This looks great. I have some minor comments regarding UX when throwing exceptions.

rsampaths16 commented on docs/tutorials/flow-task-config.md:
````diff
@@ -147,6 +167,29 @@ def my_flow():
     my_task(name="marvin", date=datetime.datetime.utcnow())
 ```
 
+Additionally this setting also accepts a function that returns a string to be used for the task_run_name
````
done

rsampaths16 commented on tests/test_flows.py:
```diff
@@ -156,11 +173,36 @@ def my_flow():
         assert my_flow.flow_run_name == "hi"
 
+    def test_flow_decorator_initializes_with_callable_flow_run_name(self):
+        # TODO: We should cover initialization with a task runner once introduced
```
removed

rsampaths16 commented on tests/test_flows.py:
```diff
@@ -156,11 +173,36 @@ def my_flow():
 
+    def test_flow_decorator_initializes_with_callable_flow_run_name(self):
+        # TODO: We should cover initialization with a task runner once introduced
+        @flow(name="foo", version="B", flow_run_name=lambda: "hi")
```
dropped

rsampaths16 commented:
Hey @madkinsz - here are the changes since the last review:
* added documentation for using `prefect.runtime` within `flow_run_name` and `task_run_name` callable
* addressed nitpicks in documentation
* made exception messages more verbose
* changed `RuntimeError`s to `TypeError`s in `prefect.engine`
* addressed nitpicks in tests
* fixed issue where `flow_run_name` is reset on retries. ( followed the same pattern used in `task_run_name`. added missing tests too ).

do let me know your thoughts on it

rsampaths16 commented:
Hi @madkinsz - I've verified that the tests failing aren't related to the current pr. Could you please check and unblock any workflows that could've been blocked by this. Thanks.

ref: https://github.com/PrefectHQ/prefect/pull/9239

rsampaths16 commented:
Hi @madkinsz - thanks for unblocking the tests via https://github.com/PrefectHQ/prefect/pull/9241!!

madkinsz reviewed (approved):
Thanks for your patience!

rsampaths16 commented:
Awesome :)

batuhanfaik commented:
Current implementation :)

```python
import datetime
from prefect import flow

def generate_flow_run_name():
    flow_run = prefect.runtime.flow_run
    date = datetime.datetime.utcnow()
    name = "anon" if not flow_run.name else str(flow_run.name)
    return f"{name}-on-{date:%A}"

@flow(flow_run_name=generate_flow_run_name)
def my_flow(name: str):
   pass

# creates a flow run called 'marvin-on-Thursday'
my_flow(name="marvin")
```
